### Візьміть датасет movielens і побудуйте модель матричної факторизації. У даній бібліотеці він має назву SVD. Підберіть найкращі параметри за допомогою крос-валідації, також поекспериментуйте з іншими алгоритмами розрахунків (SVD++, NMF) і оберіть той, який буде оптимальним.

In [2]:
pip install surprise

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 772.0/772.0 kB 10.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.3-cp310-cp310-linux_x86_64.whl size=3156236 sha256=d6fcca61d70e92fe2eb90b8b251072e681f3b12d025f8b50e7388ffa43ef7501
  Stored in directory: /root/.cache/pip/wheels/a5/ca/a8/4e28def53797fdc4363ca4af740db15a9c2f1595ebc51fb445
Successfully built scikit-surprise


In [3]:
from surprise import Dataset, Reader
from surprise.model_selection import cross_validate, GridSearchCV
from surprise.prediction_algorithms import SVD, SVDpp, NMF

In [8]:
# Завантаження данних з датасету movielens
data = Dataset.load_builtin('ml-100k')

# Створення об'єкту Reader для зчитування даних
reader = Reader(line_format='user item rating timestamp', sep='\t')

# Перетворення данних до формату, зрозумілого бібліотеці Surprise
data = Dataset.load_from_file('/content/drive/MyDrive/Data_science/hw_07/ml-100k/u.data', reader=reader)

In [9]:
# Створення списку алгоритмів
algorithms = [SVD(), SVDpp(), NMF()]

In [15]:
for algo in algorithms:
    # Використання крос-валідації для оцінки алгоритму
    results = cross_validate(algo, data, measures=['RMSE'], cv=3, verbose=True)

    # Виведення середнього значення RMSE на крос-валідації
    print(f'RMSE for {algo}: {results["test_rmse"].mean()}')

Evaluating RMSE of algorithm SVD on 3 split(s).

                  Fold 1  Fold 2  Fold 3  Mean    Std     
RMSE (testset)    0.9483  0.9491  0.9385  0.9453  0.0048  
Fit time          1.13    1.09    1.08    1.10    0.02    
Test time         0.38    0.36    0.30    0.35    0.04    
RMSE for <surprise.prediction_algorithms.matrix_factorization.SVD object at 0x7ab61476beb0>: 0.9452928743040397
Evaluating RMSE of algorithm SVDpp on 3 split(s).

                  Fold 1  Fold 2  Fold 3  Mean    Std     
RMSE (testset)    0.9326  0.9259  0.9240  0.9275  0.0037  
Fit time          18.68   18.49   18.48   18.55   0.09    
Test time         7.55    7.98    7.70    7.74    0.18    
RMSE for <surprise.prediction_algorithms.matrix_factorization.SVDpp object at 0x7ab6147698a0>: 0.9275058036805541
Evaluating RMSE of algorithm NMF on 3 split(s).

                  Fold 1  Fold 2  Fold 3  Mean    Std     
RMSE (testset)    0.9680  0.9724  0.9799  0.9734  0.0049  
Fit time          1.96    1.70    1

### За результатами оцінки RMSE, алгоритм SVDpp має найменше середнє значення RMSE (0.9275), що свідчить про його кращу точність у порівнянні з SVD (0.9452) і NMF (0.9734). Таким чином, алгоритм SVDpp є найкращим вибором серед цих трьох для побудови моделі матричної факторизації для рекомендаційної системи на даних з movielens.